crates/shared/src/price_estimation/oneinch.rs

```diff
@@ -1,15 +1,22 @@
 use crate::{
-    oneinch_api::{OneInchClient, ProtocolCache, RestResponse, SellOrderQuoteQuery},
+    oneinch_api::{
+        OneInchClient, ProtocolCache, RestResponse, SellOrderQuote, SellOrderQuoteQuery,
+    },
     price_estimation::{
         gas, Estimate, PriceEstimateResult, PriceEstimating, PriceEstimationError, Query,
     },
+    request_sharing::RequestSharing,
 };
-use futures::StreamExt;
+use futures::{future::BoxFuture, FutureExt, StreamExt};
 use model::order::OrderKind;
 use std::sync::Arc;
 
 pub struct OneInchPriceEstimator {
     api: Arc<dyn OneInchClient>,
+    sharing: RequestSharing<
+        Query,
+        BoxFuture<'static, Result<RestResponse<SellOrderQuote>, PriceEstimationError>>,
```

Comment:
The return type here is what OneInchClient returns for the request we want to share. There is no need to share the get_allowed_protocols call below because that is already cached and independent of the specific query.
The error type had to be adjusted to PriceEstimationError because the future's output must be clonable.